migrations/account_type/migration.go

```diff
@@ -0,0 +1,237 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package account_type
+
+import (
+	"github.com/onflow/cadence/migrations"
+	"github.com/onflow/cadence/runtime"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/interpreter"
+	"github.com/onflow/cadence/runtime/sema"
+)
+
+type AccountTypeMigration struct {
+	storage     *runtime.Storage
+	interpreter *interpreter.Interpreter
+}
+
+func NewAccountTypeMigration(
+	interpreter *interpreter.Interpreter,
+	storage *runtime.Storage,
+) *AccountTypeMigration {
+	return &AccountTypeMigration{
+		storage:     storage,
+		interpreter: interpreter,
+	}
+}
+
+func (m *AccountTypeMigration) Migrate(
+	addressIterator migrations.AddressIterator,
+	reporter migrations.Reporter,
+) {
+	for {
+		address := addressIterator.NextAddress()
+		if address == common.ZeroAddress {
+			break
+		}
+
+		m.migrateTypeValuesInAccount(
+			address,
+			reporter,
+		)
+	}
+
+	err := m.storage.Commit(m.interpreter, false)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// migrateTypeValuesInAccount migrates `AuthAccount` and `PublicAccount` types in a given account
+// to the account reference type (&Account).
+func (m *AccountTypeMigration) migrateTypeValuesInAccount(
+	address common.Address,
+	reporter migrations.Reporter,
+) {
+
+	accountStorage := migrations.NewAccountStorage(m.storage, address)
+
+	accountStorage.ForEachValue(
+		m.interpreter,
+		common.AllPathDomains,
+		m.migrateValue,
+		reporter,
+	)
+}
+
+func (m *AccountTypeMigration) migrateValue(value interpreter.Value) (newValue interpreter.Value, updatedInPlace bool) {
+	return migrations.MigrateNestedValue(m.interpreter, value, m.migrateTypeValue)
```

Comment:
The migration also needs to handle other values in which static types occur, e.g. dictionaries, arrays, and capabilities.

Links (`PathLinkValue`) should probably also be migrated? 🤔 I assume the migration will run before the Cap Cons migration?

Reply:
The account type is not storable, so an array/dictionary with an inner type (or nested) as an account type cannot exist in the storage in theory, right?

Reply:
> Links (PathLinkValue) should probably also be migrated? 🤔 I assume the migration will run before the Cap Cons migration?

In stable-cadence branch, I think we don't have `PathLinkValue`. Also, the Capability values are the ID-based ones (not the old path-based ones). So to avoid having to bring back those values, which may have been already done in the CapCpn migration, It might be good/easier to run the other migrations after the CapCon migration, WDYT?

Reply:
> The account type is not storable, so an array/dictionary with an inner type (or nested) as an account type cannot exist in the storage in theory, right?

I guess we can skip those containers, just worried we'll miss something.

> It might be good/easier to run the other migrations after the CapCon migration.

We can try running this migration after the Cap Cons migration, but in that case, the capability controllers (borrow types) might need to be migrated?